src/firewall/core/fw_policy.py

```diff
@@ -40,20 +40,44 @@ def new_transaction(self):
 
     # policies
 
-    def get_policies_not_derived_from_zone(self):
-        return sorted(
-            (p.name for p in self._policies.values() if not p.derived_from_zone)
-        )
-
-    def get_active_policies_not_derived_from_zone(self):
-        active_policies = []
-        for policy in self.get_policies_not_derived_from_zone():
-            p_obj = self.get_policy(policy)
-            if (set(p_obj.ingress_zones) & set(self._fw.zone.get_active_zones() + ["HOST", "ANY"])) and \
-               (set(p_obj.egress_zones)  & set(self._fw.zone.get_active_zones() + ["HOST", "ANY"])):
-                active_policies.append(policy)
-
-        return active_policies
+    def _policy_get_active_zones_set(self):
+        return set(self._fw.zone.get_active_zones() + ["HOST", "ANY"])
+
+    def policy_is_active(self, policy, active_zones=None):
+        if policy is None:
+            return False
+        if policy.derived_from_zone:
+            return False
+        if active_zones is None:
+            active_zones = self._policy_get_active_zones_set()
+        if active_zones.isdisjoint(policy.ingress_zones):
+            return False
+        if active_zones.isdisjoint(policy.egress_zones):
+            return False
+        return True
+
+    def get_policies(
+        self,
+        *,
+        require_not_derived_from_zone=True,
+        require_active=False,
+        sort=True,
+    ):
+        lst = self._policies.values()
+
+        # "require_active" implies also "require_not_derived_from_zone". Hence the if-elif.
+        if require_active:
+            active_zones = self._policy_get_active_zones_set()
+            lst = (p for p in lst if self.policy_is_active(p, active_zones))
+        elif require_not_derived_from_zone:
+            lst = (p for p in lst if not p.derived_from_zone)
+
+        lst = (p.name for p in lst)
+
+        lst = list(lst)
+        if sort:
+            lst.sort()
+        return lst
```

Comment:
previously we have:
- `get_policies()`, this returns all zones. However,
  * called by `Firewall` doing an O(n) check, it should call `FirewallZone.get_zone()` to check whether a zone exists.
  * called by `FirewallZone.{get_policies_not_derived_from_zone,apply_policies}()`, both only want user-created policies (`not p_obj.derived_from_zone`).
- `get_policies_not_derived_from_zone()`, which is used and useful. But the implementation previously called `get_policies()` to get all names, then it did a lookup by name again to find `p_obj.derived_from_zone`.
- `get_active_policies_not_derived_from_zone()` which filtered the list further, but most callers actually just want to check whether the policy is active, and don't care for the entire list.

The branch
- adds a `policy_is_active()` check, so that most callers of `get_active_policies_not_derived_from_zone()` don't need the full list.
- it unifies all three get-policies functions. They behave basically the same anyway, the difference is how they filter the arguments.


the last 3 (actually controversial) patches change that `get_policies()` returns objects instead of names. With that, it would make more sense to have separate `get-policies()` functions that apply filtering afterwards on the objects. But without those patches, `get_policies()` returns (sorted) names, and to filter them, we need to lookup the objects again. That's wasteful, but more importantly it shows that the filtering at the wrong place.

Reply:
I dropped the last 4 patches from this PR.